### PR TITLE
Tweak: Theme settings page [ED-11811]

### PR DESCRIPTION
**Pull request by rami-elementor**

### CODE REVIEW

**Review by KingYes — CHANGES_REQUESTED** [review]

File: functions.php [hunk 1]
```diff
@@ -184,6 +184,19 @@ function hello_elementor_add_description_meta_tag() {
...
 
+if ( ! function_exists( 'hello_elementor_settings_panel' ) ) {
+	/**
+	 * Add theme settings panel.
+	 *
+	 * @return void
+	 */
+	function hello_elementor_settings_panel() {
+		include_once get_template_directory() . '/includes/settings/settings-panel.php';
+		new HelloElementor\Includes\Settings\Settings_Panel();
```
Comment (KingYes) [inline functions.php:195]: @rami-elementor Maybe we should thinking about covert it to static methods?

File: includes/settings/settings-panel.php [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+class Settings_Panel {
+
+	public $menu_hook = '';
+
+	public $settings_group = 'hello_elementor_settings';
+
+	public function __construct() {
+
+		add_action( 'admin_menu', [ $this, 'register_settings_page' ], 10 );
```
Comment (KingYes) [inline includes/settings/settings-panel.php:17]: You are already on the `init` hook.

File: includes/settings/settings-panel.php [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+		add_action( 'init', [ $this, 'register_settings' ], 10 );
```
Comment (KingYes) [inline includes/settings/settings-panel.php:18]: Why did you use `10` params? It's a default value.

**Review by KingYes — CHANGES_REQUESTED** [review]

File: functions.php (hunk 1 above)
Comment (rami-elementor) [reply functions.php:195]: It's not a class method, it's a stand alone function triggered by the `init` hook.

File: includes/settings/settings-panel.php (hunk 3 above)
Comment (rami-elementor) [reply includes/settings/settings-panel.php]: Fixed. Thank you.

File: includes/settings/settings-panel.php (hunk 2 above)
Comment (rami-elementor) [reply includes/settings/settings-panel.php]: Fixed. Good catch!

File: assets/dev/js/admin/pages/main-page.js [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+				console.error( error );
+			}
+		};
+
+		if ( hasLoaded ) {
+			return;
+		}
+
+		// eslint-disable-next-line no-console
+		fetchSettings().catch( console.error );
```
[the remark this answers is not in the document]
Comment (rami-elementor) [reply assets/dev/js/admin/pages/main-page.js:127]: @KingYes  do we really need an additional `catch`? we already have a `try` `catch` inside the `fetchSettings` const (line 118).

File: assets/dev/js/admin/pages/main-page.js [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	useEffect( () => {
+		const fetchSettings = async () => {
+			try {
+				await api.loadPromise;
+				const settings = new api.models.Settings();
+				const response = await settings.fetch();
+				setSettingsData( {
+					_description_meta_tag: response[ settingsPrefix + '_description_meta_tag' ],
+					_skip_link: response[ settingsPrefix + '_skip_link' ],
+					_page_title: response[ settingsPrefix + '_page_title' ],
```
Comment (bainternet) [inline assets/dev/js/admin/pages/main-page.js:103]: please move setting names to a single const object and reference from there
```JS
const SETTINGS = {
 DESCRIPTION_META_TAG: 'description_meta_tag',
};
```

File: assets/dev/js/admin/pages/main-page.js [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+
+						<ToggleControl
+							label={ __( 'Disable description meta tag', 'hello-elementor' ) }
+							help={
+								sprintf(
+									/* translators: %s: The <head> tag. */
+									__( 'A meta tag that contains the post/page excerpt, in the %s.', 'hello-elementor' ),
+									'<head>',
+								)
+							}
```
Comment (bainternet) [inline assets/dev/js/admin/pages/main-page.js:167]: please split panels to separate component files

**Review by bainternet — CHANGES_REQUESTED** [review]

File: includes/settings-functions.php [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+
+	$option = get_option( $settings_group . '_skip_link' );
+	if ( isset( $option ) && ( 'true' === $option ) ) {
+		add_filter( 'hello_elementor_enable_skip_link', '__return_false' );
+	}
+
+	$option = get_option( $settings_group . '_page_title' );
+	if ( isset( $option ) && ( 'true' === $option ) ) {
+		add_filter( 'hello_elementor_page_title', '__return_false' );
+	}
```
[the remark this answers is not in the document]
Comment (rami-elementor) [reply includes/settings-functions.php]: Done.

File: assets/dev/js/admin/pages/main-page.js (hunk 6 above)
Comment (rami-elementor) [reply assets/dev/js/admin/pages/main-page.js]: Done.

File: assets/dev/js/admin/pages/main-page.js (hunk 5 above)
Comment (rami-elementor) [reply assets/dev/js/admin/pages/main-page.js]: Done.

File: includes/settings-functions.php [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+/**
+ * Run a tweek only if the user requested it.
+ */
+function hello_elementor_do_tweak( $setting, $tweak_callback ) {
+
+	$option = get_option( $setting );
+	if ( isset( $option ) && ( 'true' === $option ) ) {
+		$tweak_callback();
+	}
```
Comment (bainternet) [inline includes/settings-functions.php:138]:
```suggestion
	if ( isset( $option ) && ( 'true' === $option ) && is_callable( $tweak_callback ) ) {
		$tweak_callback();
	}
```

Comment (bainternet) [reply includes/settings-functions.php:138]: always make sure you can call it

**Review by bainternet — CHANGES_REQUESTED** [review]
Looks better,

* left a suggestion for `is_callable`
* I would also move all PHP settings in to constants but not a must.
* The react is much more readable now, I would go even farther and move all settings logic to a custom hook.

File: includes/settings-functions.php (hunk 8 above)
Comment (rami-elementor) [reply includes/settings-functions.php]: Done.

**Review by KingYes — APPROVED** [review]

**Pull request merged**